lib/flex_algo.h

```diff
@@ -0,0 +1,172 @@
+/*
+ * Flexible Algorithm definitions
+ * Copyright (C) 2022  Hiroki Shirokura, LINE Corporation
+ * Copyright (C) 2022  Masakazu Asama
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef _FRR_FLEX_ALGO_H
+#define _FRR_FLEX_ALGO_H
+
+#include "linklist.h"
+#include "prefix.h"
+#include "bitfield.h"
+#include "segment_routing.h"
+
+#define ADMIN_GROUP_PRINT_MAX_SIZE 2048
+#define AFFINITY_NAME_SIZE 256
+#define FLEX_ALGO_PRIO_DEFAULT 128
+#define EXT_ADMIN_GROUP_MAX_POSITIONS 1024
+
+#define CALC_TYPE_SPF 0
+
+/* flex-algo definition flags */
+
+/* M-flag (aka. prefix-metric)
+ * Flex-Algorithm specific prefix and ASBR metric MUST be used
+ */
+#define FAD_FLAG_M 0x80
+
+/*
+ * Metric Type values from draft-ietf-lsr-flex-algo-19
+ */
+enum flex_algo_metric_type {
+	MT_IGP = 0,
+	MT_MIN_UNI_LINK_DELAY = 1,
+	MT_TE_DEFAULT = 2,
+};
+
+struct affinity_map {
+	char name[AFFINITY_NAME_SIZE];
+	uint16_t bit_position;
+};
+
+struct affinity_maps {
+	struct list *maps;
+};
+
+struct admin_group {
+	bitfield_t bitmap;
+};
+
+struct flex_algo {
+	uint8_t algorithm;
+	enum flex_algo_metric_type metric_type;
+	uint8_t calc_type;
+	uint8_t priority;
+	uint8_t flags;
+
+#define FLEX_ALGO_SR_MPLS 0x01
+#define FLEX_ALGO_SRV6 0x02
+#define FLEX_ALGO_IP 0x04
+	uint8_t dataplanes;
+
+	/* index is the FAD sub-tlv type.
+	 * Entry is true if an unsupported sub-TLV type is received.
+	 * Currently, Exclude SRLG Sub-TLV is unsupported.
+	 * Future sub-TLVs are considered as unsupported.
+	 * */
+	bool unsupported_subtlv[256];
```

Comment:
I don't see the use of this. You either support a sub-TLV type or you just ignore them. Why do you need to keep track of a unsupported TLV type is present or not in the flex-also definition? You are not even storing the unsupported TLV-type contents here.

Reply:
draft-ietf-lsr-flex-algo-25 specifies:
```
If a node is configured to participate in a particular Flexible-
Algorithm, but (...) the selected Flex-Algorithm definition includes
(...) Sub-TLV that is not supported by the node, it MUST stop
participating in such Flexible-Algorithm.
```
It considers a received Flex-Algo Definition with an unknown Sub-TLV as
unsupported so that the local router stop participating in specified
Flexible-Algorithm. Future unsupported Sub-TLVs MUST stop the participation in the Flex-algo.
The list is intended to display which unsupported Sub-TLV has been received in show isis flex-algo

Reply:
So you are suggesting storing a Flex-Algo den received from remote routers in the Flex-Algo Db too. I have the same opinion as @odd22 that such definitions learnt should be in the LinkState library.

You are contradicting your own comment above (relayed to Dataplane config) that the Flex-Algo-Db is not for definition but for storing configurations.

The FADs learnt over wire should not be stored in Flex-Algo-Db and should be in Linkstate along with the ones advertised by the local node (just like local links and prefixes).

Reply:
@pushpasis 
> So you are suggesting storing a Flex-Algo den received from remote routers in the Flex-Algo Db too. I have the same opinion as @odd22 that such definitions learnt should be in the LinkState library.
I talk with @odd22 . Since I was not aware of the LinkState library requirement, it will not be done in this pull request. 

> You are contradicting your own comment above (relayed to Dataplane config) that the Flex-Algo-Db is not for definition but for storing configurations.
I did not say that my struct was about definition. My concern was that your structure `struct flex_algo_defn` in https://github.com/FRRouting/frr/pull/11849/ has definition in naming but it is not all about definition but configuration.

> The FADs learnt over wire should not be stored in Flex-Algo-Db and should be in Linkstate along with the ones advertised by the local node (just like local links and prefixes).
I am not storing learnt FAD in what you call the Flex-Algo Db. They stay in lsdb. However, if I receive an unsupported definition subtlv, I store the info in Flex-Algo Db.